### PR TITLE
Roll your own Docker containers, part 1

### PR DESCRIPTION
**Pull request by rdormer**

### CODE REVIEW

File: _posts/2019-08-15-docker-containers-pt-1.markdown [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+categories: ["DevOps", "docker", "tutorials"]
+author: "rdormer"
+---
+
+In my [last article](https://www.ombulabs.com/blog/devops/docker/tutorials/gentle-intro-to-docker.html), I gave a brief tour of what Docker is, and basic usage. Once you've got your feet under you with basic usage, you'll reach the point where you want to start making your own containers. You'll certainly need to know how to roll your own if you want to use Docker in your own infrastructure.
+
+## Container basics
+
+Broadly speaking, there are two types of Docker images - *base images*, and images that are built on top of them. Most of the time, when you're making an image of your own, you'll be inheriting from a base (or parent) image, and tweaking it by layering changes over top of it.  Layering is exactly what you're dong, by the way - as you execute commands and copy data into a container, all of the changes you're making are written to a *writable layer* (also known as the *container layer*). There's a lot more to this particular subject - see [storage drivers](https://docs.docker.com/storage/storagedriver/) and [volumes](https://docs.docker.com/storage/volumes/) for a lot more information.
```
Comment (lubc) [inline _posts/2019-08-15-docker-containers-pt-1.markdown]: typo:
```suggestion
Broadly speaking, there are two types of Docker images - *base images*, and images that are built on top of them. Most of the time, when you're making an image of your own, you'll be inheriting from a base (or parent) image, and tweaking it by layering changes over top of it.  Layering is exactly what you're doing, by the way - as you execute commands and copy data into a container, all of the changes you're making are written to a *writable layer* (also known as the *container layer*). There's a lot more to this particular subject - see [storage drivers](https://docs.docker.com/storage/storagedriver/) and [volumes](https://docs.docker.com/storage/volumes/) for a lot more information.
```

File: _posts/2019-08-15-docker-containers-pt-1.markdown [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+`FROM debian`
+
+Which specifies that you'll be building your image on top of [Debian](https://hub.docker.com/_/debian).  Usually, the image that you'll be building will be, at least partially, a copy of some application that you're running.  You can create an image of this application by putting your Dockerfile in a directory containing a clean installation of your application. All of the commands in your Dockerfile will be relative to this directory, which Docker calls the *context*.  Add the [following command](https://docs.docker.com/engine/reference/builder/#copy) after the `FROM` command:
+
+`COPY . /myapp`
+
+If you have [any other commands that need to be run](https://docs.docker.com/engine/reference/builder/#run), you can use the `CMD` command.  Let's use it to install [RVM](https://rvm.io/):
+
+`CMD apt-get update`
```
Comment (bronzdoc) [inline _posts/2019-08-15-docker-containers-pt-1.markdown]: `CMD` is not used like that.

You would like to replace that with `RUN`

See: https://goinbigdata.com/docker-run-vs-cmd-vs-entrypoint/

File: _posts/2019-08-15-docker-containers-pt-1.markdown [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+`FROM debian`
+`CMD apt-get update`
+`CMD apt-get install curl`
+`CMD curl -sSL https://get.rvm.io | bash -s stable --ruby`
+`COPY . /myapp`
+
+Run this command:
+
+`docker build -t myfirstcontainer .`
```
Comment (bronzdoc) [inline _posts/2019-08-15-docker-containers-pt-1.markdown]: Maybe specify that if no tag is given the default `latest` tag will be created?
```
-t, --tag list                Name and optionally a tag in the 'name:tag' format
```

File: _posts/2019-08-15-docker-containers-pt-1.markdown (hunk 1 above)
Comment (rdormer) [reply _posts/2019-08-15-docker-containers-pt-1.markdown]: Hah, good catch.  Definitely not a typo we should let go out :)

File: _posts/2019-08-15-docker-containers-pt-1.markdown (hunk 3 above)
Comment (rdormer) [reply _posts/2019-08-15-docker-containers-pt-1.markdown]: @bronzdoc speaking as a beginner - it's best not to overwhelm with too much info up front.  Start out as simply as possible.  Details can be filled in later.

**Pull request merged**